plugins/woocommerce-blocks/assets/js/base/components/cart-checkout/form/validate-state.ts

```diff
@@ -0,0 +1,63 @@
+/**
+ * External dependencies
+ */
+import { dispatch, select } from '@wordpress/data';
+import { ShippingAddress } from '@woocommerce/settings';
+import { VALIDATION_STORE_KEY } from '@woocommerce/block-data';
+import { __ } from '@wordpress/i18n';
+
+function previousString( initialValue?: string ) {
+	let lastValue = initialValue;
+
+	function track( value: string ) {
+		const currentValue = lastValue;
+
+		lastValue = value;
+
+		// Return the previous value
+		return currentValue;
+	}
+
+	return track;
+}
+
+const lastValue = previousString( '' );
+
+export const validateState = (
+	addressType: string,
+	values: ShippingAddress,
+	isRequired: boolean
+) => {
+	const validationErrorId = `${ addressType }_state`;
+	const hasValidationError =
+		select( VALIDATION_STORE_KEY ).getValidationError( validationErrorId );
+
+	const countryChanged = lastValue( values.country ) !== values.country;
+
+	if ( hasValidationError ) {
```

Comment:
A semi-complex set of states to determine if we should show a validation error for the state.

If we choose a country that has a required state then we "set" the validation error (which puts it hidden in state). If the user tries to submit with the same country we show the error.

If the country is changed to a non-required state we remove the validation, if its changed to another country with required state the validation is set but hidden.

Reply:
I'm wondering why we need to keep track of previous countries and not simply check the current state value and requirement? Is this because unsetting can be done via user (show error) or via country change (hide error)?

Reply:
> Is this because unsetting can be done via user (show error) or via country change (hide error)?
@senadir Yes exactly.

I needed to distinguish the valid situation where the country was changed to avoid showing the error.

Reply:
I've adjusted the logic of this because we also need to determine if other address fields change. If they changed we don't want to show the state validation (ie if you edit "City", we don't need to immediately inform you that state is required, it doesn't feel necessary). We only want to show it if the user tried to submit the form without it.

The only thing we don't have is an "on blur" validation, but in testing (imo) I don't think this impacted ux. Of course this is all just opinion and I'm open to changing that :)